### PR TITLE
new resource "azurerm_automation_connection", azurerm_automation_connection_certificate`, `azurerm_automation_connection_classic_certificate` and `azurerm_automation_connection_service_pricipal`

### LINKED ISSUE #3881: Azure automation_account connection and certificate objects
_This issue was originally opened by @bikashkarmakar as hashicorp/terraform#22141. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    
Hi 
We use automation account with a service principal (created with certifiacte -> az ad sp create-for-rbac --name ServicePrincipalName01 --create-cert ). We need to associate the automation account connection (of type AzureServicePrincipal) and certificate (the Service Principal certificate).

As per https://www.terraform.io/docs/providers/azurerm/r/automation_account.html , there is no object for connection and certificate. Can you please advise if this is supported ?

### PR DESCRIPTION
**Pull request by njuCZ**
automation account API does not provider `run as account` interface, users need to configure `service principal`, `automation certificate` and `automation connection` manually. 
This is the last PR about the flow.

`azurerm_automation_connection_certificate`, `azurerm_automation_connection_classic_certificate`, `azurerm_automation_connection_service_pricipal` coresponds to 
resource `azurerm_automation_connection_certificate` coresponds to builtin types: `Azure`, `AzureClassicCertificate` , `AzureServicePrincipal`.

`azurerm_automation_connection` is designed for connecting user's custom connection type. (user could add new types).

https://docs.microsoft.com/en-us/azure/automation/automation-connections#connection-types
https://docs.microsoft.com/en-us/azure/automation/automation-connections#add-a-connection-type

image

relevant issues
#4431
#3881

(fixes #3881)

### CODE REVIEW

File: website/docs/r/automation_connection.html.markdown [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+* `type` - (Required) The type of the Connection - can be either builtin type `Azure`, `AzureClassicCertificate` `AzureServicePrincipal` or user custom types. Changing this forces a new resource to be created.
+
+* `values` - (Optional) A mapping of key value pairs passed to the connection. Different `type` needs different parameters in the `values`. Builtin types have required field values as below:
+
+  - `Azure`: parameters `AutomationCertificateName` and `SubscriptionID`.
+
+  - `AzureClassicCertificate`: parameters `SubscriptionName`, `SubscriptionId` and `CertificateAssetName`.
+
+  - `AzureServicePrincipal`: parameters `ApplicationId`, `CertificateThumbprint`, `SubscriptionId` and `TenantId`.
```
Comment (katbyte) [inline website/docs/r/automation_connection.html.markdown:70]: I think it might be better to have 3 different resources with these properties exposed

**Comment by njuCZ** [comment]
@katbyte because the docs says user could add a custom connection type and required parameters. So if we split them into three resources, I am afraid this feature could not be supported.

**Comment by katbyte** [comment]
@njuCZ - then could we create the three strongly typed resources and then this one can be used for user defined ones?

**Comment by njuCZ** [comment]
@katbyte ok, thanks for your suggestion

**Comment by njuCZ** [comment]
sorry for busy other things yesterday, I will update this PR as soon as possible

**Comment by njuCZ** [comment]
@katbyte I have updated the PR, please have a look

**Comment by lazywinadmin** [comment]
Any update on when this will be released ? @katbyte 
It would be great to be able to create connection resources within Automation Account
Thanks

File: azurerm/internal/services/automation/tests/automation_connection_certificate_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+func testAccAzureRMAutomationConnectionCertificate_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/automation/tests/automation_connection_certificate_resource_test.go:159]: Could we make this:
```suggestion
  name     = "acctestRG-auto-%d"
```

File: azurerm/internal/services/automation/tests/automation_connection_classic_certificate_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+func testAccAzureRMAutomationConnectionClassicCertificate_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Comment (katbyte) [inline azurerm/internal/services/automation/tests/automation_connection_classic_certificate_resource_test.go:162]: Could we make this:
```suggestion
  name     = "acctestRG-auto-%d"
```

File: website/docs/r/automation_connection.html.markdown [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of the Connection. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the resource group in which the Connection is created. Changing this forces a new resource to be created.
+
+* `automation_account_name` - (Required) The name of the automation account in which the Connection is created. Changing this forces a new resource to be created.
+
+* `type` - (Required) The type of the Connection - can be either builtin type `Azure`, `AzureClassicCertificate` `AzureServicePrincipal` or user custom types. Changing this forces a new resource to be created.
```
Comment (katbyte) [inline website/docs/r/automation_connection.html.markdown:62]:
```suggestion
* `type` - (Required) The type of the Connection - can be either builtin type such as `Azure`, `AzureClassicCertificate`, and `AzureServicePrincipal`, or a user defined types. Changing this forces a new resource to be created.
```

File: website/docs/r/automation_connection.html.markdown [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+* `update` - (Defaults to 30 minutes) Used when updating the Automation Connection.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Automation Connection.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Automation Connection.
+
+## Import
+
+Automation Connection can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_automation_connection.conn1 /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Automation/automationAccounts/account1/connections/conn1
```
Comment (katbyte) [inline website/docs/r/automation_connection.html.markdown:94]:
```suggestion
terraform import azurerm_automation_connection.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Automation/automationAccounts/account1/connections/conn1
```

File: website/docs/r/automation_connection_certificate.html.markdown [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+subcategory: "Automation"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_automation_connection_certificate"
+description: |-
+  Manages an Automation Connection with type `Azure`.
+---
+
+# azurerm_automation_connection_certificate
+
+Manages a Automation Connection with type `Azure`.
```
Comment (katbyte) [inline website/docs/r/automation_connection_certificate.html.markdown:11]:
```suggestion
Manages an Automation Connection with type `Azure`.
```

File: website/docs/r/automation_connection_certificate.html.markdown [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+* `update` - (Defaults to 30 minutes) Used when updating the Automation Connection.
+* `read` - (Defaults to 5 minutes) Used when retrieving the Automation Connection.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Automation Connection.
+
+## Import
+
+Automation Connection can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_automation_connection_certificate.conn1 /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Automation/automationAccounts/account1/connections/conn1
```
Comment (katbyte) [inline website/docs/r/automation_connection_certificate.html.markdown:89]:
```suggestion
terraform import azurerm_automation_connection_certificate.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Automation/automationAccounts/account1/connections/conn1
```

File: website/docs/r/automation_connection_classic_certificate.html.markdown [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+subcategory: "Automation"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_automation_connection_classic_certificate"
+description: |-
+  Manages a Automation Connection with type `AzureClassicCertificate`.
+---
+
+# azurerm_automation_connection_classic_certificate
+
+Manages a Automation Connection with type `AzureClassicCertificate`.
```
Comment (katbyte) [inline website/docs/r/automation_connection_classic_certificate.html.markdown:11]:
```suggestion
Manages an Automation Connection with type `AzureClassicCertificate`.
```

File: website/docs/r/automation_connection_classic_certificate.html.markdown [hunk 9]
```diff
@@ -0,0 +1,86 @@
+---
+subcategory: "Automation"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_automation_connection_classic_certificate"
+description: |-
+  Manages a Automation Connection with type `AzureClassicCertificate`.
```
Comment (katbyte) [inline website/docs/r/automation_connection_classic_certificate.html.markdown:6]:
```suggestion
  Manages an Automation Connection with type `AzureClassicCertificate`.
```

File: website/docs/r/automation_connection_service_principal.html.markdown [hunk 10]
```diff
@@ -0,0 +1,89 @@
+---
+subcategory: "Automation"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_automation_connection_service_principal"
+description: |-
+  Manages a Automation Connection with type `AzureServicePrincipal`.
```
Comment (katbyte) [inline website/docs/r/automation_connection_service_principal.html.markdown:6]:
```suggestion
  Manages an Automation Connection with type `AzureServicePrincipal`.
```

File: website/docs/r/automation_connection_service_principal.html.markdown [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+---
+
+# azurerm_automation_connection_service_principal
+
+Manages a Automation Connection with type `AzureServicePrincipal`.
```
Comment (katbyte) [inline website/docs/r/automation_connection_service_principal.html.markdown:11]:
```suggestion
Manages an Automation Connection with type `AzureServicePrincipal`.
```

**Pull request merged**